### ISSUE TITLE
javascript: copy-pasting code containing switch statement indents weirdly

### ISSUE THREAD

**Issue by AgentME** [atom/atom#4648]
If you copy this code:

```
function test(a) {
  switch(a) {
    case 1:
      console.log('one');
      break;
    case 2:
      console.log('two');
      break;
    default:
      console.log('uh oh');
  }
}
```

and then try to paste it into a javascript file in Atom, you get this:

```
function test(a) {
  switch(a) {
    case 1:
      console.log('one');
      break;
      case 2:
        console.log('two');
        break;
        default:
          console.log('uh oh');
        }
      }
      
```

Atom version 0.158.0

**Comment by rolandkopka** [comment]
Same with C++

https://gist.github.com/rolandkopka/9acb1972a270d8c15846
Ctrl-A, Ctrl-C, Ctrl-V the Raw of the gist into an empty cpp file. This is the result:
![screenshot from 2015-01-03 17 51 23 atom](https://cloud.githubusercontent.com/assets/7033460/5603018/795ff744-9371-11e4-986e-07f68df9c30b.png)
The blue squares are trailing spaces.

**Comment by zeke** [comment]
This fixed it for me:

>  Under settings => "auto indent on paste" -  I had to turn that off, not sure if was a new default or what.

via https://twitter.com/mikeal/status/551087742918995968

**Comment by tannyo** [comment]
I was having problems pasting javascript code. Especially objects that contained an array of objects. The code would unindent for each object. Unchecking 'auto indent on paste' fixed this for me.

**Comment by WadeShuler** [comment]
@tannyo Doesn't unchecking 'auto indent on paste' cause the indents to not line up correctly in your code? I haven't tested it, but just judging by the name of the option, I assume your pasted in code will then not line up.

For example:

Say this is your php file:
    class MyClass {
        private function test() {
            echo 'my test';
        }
    }

Then say this is from a website (github, stackoverflow, joe's blog, or from another editor):
    private function tryMe() {
        echo 'just try me';
    }

Notice how the tryMy() function is not indented. In your class, your functions are indented. I expect, with 'auto indent on paste' off, that it would not place the function 1 indentation in, and it would be the same level of indentation as the class block, which is incorrect.

If I am wrong, then what IS the purpose of 'auto indent on paste'? If I am right, then turning off 'auto indent on paste' just removes the whole purpose of using a sophisticated editor. If I copy from N++ 3 functions worth 600 lines of code, I don't want to manually fix the indentations, and I shouldn't have to.

Those saying "fixed it for me" or "solved it" are incorrect. It is a work around or a hack, not a fix or solution. A solution, would be to fix the bug in the next release.

This is persistent for every language (javascript, php, etc) and affects comment blocks, multi-line arrays, and probably anything else that extends on multiple lines such as array().

**Comment by 50Wliu** [comment]
I believe unchecking both 'Auto-indent on paste' and 'Normalize indent on paste' will cause Atom to paste whatever code you have in without doing any indentation modifications to it.

Anyway, there's currently an open PR that is aiming to fix this (see atom/language-javascript#92).

**Comment by tannyo** [comment]
@WadeShuler Yes, I now have to manually indent when I paste code from external resources. If I move code within a module I have to manually set the indentation. However, it's much better than before when after pasting code I would have to manually indent code that I pasted within the same project that was already correctly indented. Now I just have to take the block of code I just pasted and indent or outdent it which is much better than having to go line by line to get everything correct.

It's better than taking this:

![image](https://cloud.githubusercontent.com/assets/1413384/5845563/d4993c20-a171-11e4-8045-e4c9d52e44ad.png)

which is properly indented for javascript and getting this:

![image](https://cloud.githubusercontent.com/assets/1413384/5845567/df13a7c6-a171-11e4-81f5-79df36cd17de.png)

Seems like the opposite problem @AgentME was having. My code outdents, his code indents.

When typing code it mostly works.
